src/thunderbots/software/ai/hl/stp/tactic/shoot_goal_tactic.cpp

```diff
@@ -0,0 +1,161 @@
+}
+
+void ShootGoalTactic::updateParams(const Field &field, const Team &friendly_team,
+                                   const Team &enemy_team, const Ball &ball)
+{
+    this->field         = field;
+    this->friendly_team = friendly_team;
+    this->enemy_team    = enemy_team;
+    this->ball          = ball;
+}
+
+double ShootGoalTactic::calculateRobotCost(const Robot &robot, const World &world)
+{
+    auto ball_intercept_opt =
+        Evaluation::findBestInterceptForBall(world.ball(), world.field(), robot);
+    double cost = 0;
+    if (ball_intercept_opt)
+    {
+        // If we can intercept the ball, use the distance to the intercept point.
+        // We normalize with the total field length so that robots that are within the
+        // field have a cost less than 1
+        cost = (ball_intercept_opt->first - robot.position()).len() /
+               world.field().totalLength();
+    }
+    else
+    {
+        // If we can't intercept the ball, just use the distance to the ball's current
+        // position. We normalize with the total field length so that robots that are
+        // within the field have a cost less than 1
+        cost = (world.ball().position() - robot.position()).len() /
+               world.field().totalLength();
+    }
+
+    return std::clamp<double>(cost, 0, 1);
+}
+
+bool ShootGoalTactic::hasShotAvailable() const
+{
+    return has_shot_available;
+}
+
+bool ShootGoalTactic::isEnemyAboutToStealBall() const
+{
+    // TODO: replace with get all robots except goalie?
+    for (const auto &enemy : enemy_team.getAllRobots())
+    {
+        if ((enemy.position() - ball.position()).len() < ENEMY_DANGER_DIST)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+std::optional<std::pair<Point, double>> ShootGoalTactic::getShotData() const
+{
+    auto best_shot_opt = Evaluation::calcBestShotOnEnemyGoal(field, friendly_team,
+                                                             enemy_team, ball.position());
+    if (best_shot_opt)
+    {
+        double net_percent_open =
+            Evaluation::calcShotOpenNetPercentage(field, ball.position(), *best_shot_opt);
+        return std::make_pair(best_shot_opt->first, net_percent_open);
+    }
+
+    return std::nullopt;
+}
+
+void ShootGoalTactic::shootUntilShotBlocked(KickAction &kick_action,
+                                            ChipAction &chip_action,
+                                            IntentCoroutine::push_type &yield) const
+{
+    auto shot_target = getShotData();
+    while (shot_target)
+    {
+        if (!isEnemyAboutToStealBall())
+        {
+            // If we are in the middle of committing to a shot but an enemy is about to
+            // steal the ball we chip instead to just get over the enemy. We do not adjust
+            // the point we are targeting since that may take more time to realign to, and
+            // we need to be very quick so the enemy doesn't get the ball
```

Comment:
This comment seems like it belongs in the `else` case below?

Reply:
Also is this still true if we give this tactic a chip target?

Reply:
Yup moved to else case. The chip target only applies if we aren't in the middle of committing to a shot. If we are already committing then we don't have time to adjust our aim before the enemy gets to us, so we chip in the current direction. Otherwise, we use the chip target